### PR TITLE
Cancel response stream when connection closes

### LINKED ISSUE #9068: `ReadableStream` in response body not canceled when client closes connection
### Astro Info

```block
Astro                    v3.5.2
Node                     v20.2.0
System                   macOS (x64)
Package Manager          pnpm
Output                   server
Adapter                  @astrojs/node
Integrations             @astrojs/tailwind
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

When implementing server-sent events, `cancel()` is not called when client closes the connection (e.g. closes the tab).

```ts
export async function GET() {
  const stream = new ReadableStream({
    start(controller) {
      // do stuff
    },
    cancel() {
      console.log("cancelled");
    },
  });
  return new Response(stream, {
    headers: {
      "X-Content-Type-Options": "nosniff",
      "Content-Type": "text/event-stream; charset=utf-8",
      "Transfer-Encoding": "chunked",
    },
  });
}
```

### What's the expected result?

`cancelled` is logged to the console

### Link to Minimal Reproducible Example

https://github.com/pilcrowOnPaper/astro-readable-stream-bug

### Participation

- [x] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by pilcrowOnPaper**
## Changes

- Closes #9068
- Remove unused code

## Testing

No test added, would love some directions

## Docs

None

### CODE REVIEW

**Comment by lilnasy** [comment]
Thanks for the enhancement, the code looks great! Good eye on the unused modules as well.

Do you think this behavior could be easily tested?

**Comment by pilcrowOnPaper** [comment]
@lilnasy Hm, it should be possible, though I'm not sure what's the best way to test step 4

1. Make `fetch()` request to server
2. Return response with stream
3. Abort `fetch()` request
4. Check if `cancel()` was called

**Comment by lilnasy** [comment]
What is the current unwanted behavior?

Is it that the readable stream continues to produce chunks, possibly endlessly?

**Comment by pilcrowOnPaper** [comment]
> Is it that the readable stream continues to produce chunks, possibly endlessly?

Yeah. This is especially an issue for server sent events since there's no definite end to the stream. It continues to push data until the client closes the connection

**Comment by lilnasy** [comment]
Yeah, we could use locals for that. I'll take a look.

File: .changeset/modern-ways-develop.md
Comment (lilnasy) [inline .changeset/modern-ways-develop.md:5]:
```suggestion
The node adapter now automatically cancels the `ReadableStream` of your streaming responses when the visitor disconnects, allowing you to gracefully clean up resources.
```
Just making it more descriptive. LMK if this looks helpful.

Comment (pilcrowOnPaper) [reply .changeset/modern-ways-develop.md:5]: eh, I personally find the original descriptive enough (when compared to other changesets). Gonna defer to the maintainers, I don't have a strong preference

File: packages/integrations/node/test/api-route.test.js
Comment (lilnasy) [inline packages/integrations/node/test/api-route.test.js:105]: This line might need another look. I wasn't able to mock actual disconnection which would have emitted the event realistically. I don't know enough about node to take it further.

Comment (pilcrowOnPaper) [reply packages/integrations/node/test/api-route.test.js:105]: I think this fine 

>Indicates that the response is completed, or its underlying connection was terminated prematurely (before the response completion).

https://nodejs.org/api/http.html#event-close_2

**Review by lilnasy — APPROVED** [review]
LGTM, thanks!

**Review by bluwy — APPROVED** [review]
The original PR that added the response iterator was https://github.com/withastro/astro/pull/5418, but it sounds like this PR aligns more with how Response work as a standard, so I think this is the right approach.

File: .changeset/modern-ways-develop.md
Comment (bluwy) [reply .changeset/modern-ways-develop.md:5]: I don't really have a hard preference either, but since this is only a patch, I think a short description is also fine by me.

**Comment by pilcrowOnPaper** [comment]
Is it true that the response body can be a `Buffer`? 

> Sometimes Astro sends a ReadableStream or a Buffer as a response body

**Comment by bluwy** [comment]
I think it's possible in the past when we support returning simple objects in endpoints, but now that we enforce and coerce to Response, it shouldn't apply now.

File: .changeset/modern-ways-develop.md
Comment (sarah11918) [inline .changeset/modern-ways-develop.md:5]:
```suggestion
Fixes a bug where the response stream would not cancel when the connection closed
```
(Not blocking) I can meet you halfway! What about something of this style? (Make it accurate.)

(I did like @lilnasy 's version, but can appreciate that everyone wants something shorter for a patch!)

**Pull request merged**
